openstack/imageservice/v2/images/results.go

```diff
@@ -0,0 +1,404 @@
+	if any, ok := m[k]; ok {
+		if str, ok := any.(string); ok {
+			return stringToImageVisibility(str)
+		}
+		return "", fmt.Errorf("expected string as \"%s\" value, but found: %#v", k, any)
+	}
+	return "", fmt.Errorf("expected key \"%s\" in map, but this key is not present", k)
+}
+
+func extractBoolAtKeyOptional(m map[string]interface{}, k string, ifMissing bool) (bool, error) {
+	if any, ok := m[k]; ok {
+		return asBool(any)
+	}
+	return ifMissing, nil
+
+}
+
+func extractMapStringStringAtKeyOptional(m map[string]interface{}, k string, ifMissing map[string]string) (map[string]string, error) {
+	if any, ok := m[k]; ok {
+		return asMapStringString(any)
+	}
+	return ifMissing, nil
+}
+
+func extractImage(res gophercloud.ErrResult) (*Image, error) {
+	if res.Err != nil {
+		return nil, res.Err
+	}
+
+	body, ok := res.Body.(map[string]interface{})
+	if !ok {
+		return nil, fmt.Errorf("expected map as result body, but found: %#v", res.Body)
+	}
+
+	var image Image
+
+	var err error
+
+	if image.ID, err = extractStringAtKey(body, "id"); err != nil {
+		return nil, err
+	}
+
+	if image.Name, err = extractStringAtKey(body, "name"); err != nil {
+		return nil, err
+	}
+
+	if image.Status, err = extractImageStatusAtKey(body, "status"); err != nil {
+		return nil, err
+	}
+
+	if image.Tags, err = extractStringSliceAtKey(body, "tags"); err != nil {
+		return nil, err
+	}
+
+	if image.ContainerFormat, err = extractNoneableStringAtKey(body, "container_format"); err != nil {
+		return nil, err
+	}
+
+	if image.DiskFormat, err = extractNoneableStringAtKey(body, "disk_format"); err != nil {
+		return nil, err
+	}
+
+	if image.MinDiskGigabytes, err = extractNoneableIntegerAtKey(body, "min_disk"); err != nil {
+		return nil, err
+	}
+
+	if image.MinRAMMegabytes, err = extractNoneableIntegerAtKey(body, "min_ram"); err != nil {
+		return nil, err
+	}
+
+	if image.Owner, err = extractNoneableStringAtKey(body, "owner"); err != nil {
+		return nil, err
+	}
+
+	// FIXME should this key actually be optional? Is a missing key equivalent to "protected": false ?
+	if image.Protected, err = extractBoolAtKeyOptional(body, "protected", false); err != nil {
+		return nil, err
+	}
+
+	if image.Visibility, err = extractImageVisibilityAtKey(body, "visibility"); err != nil {
+		return nil, err
+	}
+
+	if image.Checksum, err = extractNoneableStringAtKey(body, "checksum"); err != nil {
+		return nil, err
+	}
+
+	if image.SizeBytes, err = extractNoneableIntegerAtKey(body, "size"); err != nil {
+		return nil, err
+	}
+
+	if image.Metadata, err = extractMapStringStringAtKeyOptional(body, "metadata", make(map[string]string)); err != nil {
+		return nil, err
+	}
+
+	if image.Properties, err = extractMapStringStringAtKeyOptional(body, "properties", make(map[string]string)); err != nil {
+		return nil, err
+	}
+
+	return &image, nil
+}
+
+// Extract build CreateResults from imput Image
+func (c CreateResult) Extract() (*Image, error) {
```

Comment:
Instead of all those functions, try implementing something like this: https://github.com/rackspace/gophercloud/blob/master/openstack/compute/v2/servers/results.go#L19